### DISCUSSION TITLE
Slow response times on Vercel - a way to move the session API outside of Vercel while hosting the rest of the application there?

### DISCUSSION THREAD

**Discussion by avisra** [nextauthjs/next-auth#6279] [Help]
### Question 💬

If I host my entire application on Azure App Services, the session API calls (from client) are very fast - usually around 30ms. The provider is a self-hosted (in Azure) IdentityServer.

But, when my application is hosted by Vercel, that session API is hosted by Vercel's serverless functions. When hosted here, it takes 150ms to 200ms. And this is after the cold-start. During a cold start... this can vary from 8ms to 15ms.

I love everything that Vercel provides - but would like full control over the performance of my application. All of my other APIs are delivered from a Azure App Service - and they are very fast. Is there a way to externalize the session API so it is handled by an endpoint in my Azure App Services?

### How to reproduce ☕️

Deploy a NextAuth application to Vercel and compare to a NextAuth application hosted on Azure App Services

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
Hi, this is likely not a Vercel issue. do you have a reproduction of a slow endpoint?

**Comment by avisra** [comment]
@balazsorban44 I have it on my test site, but I can't put that link here publicly. I'll setup a separate reproduction for you.

The endpoint that I'm currently try to make faster is the "/api/auth/session" endpoint. On Vercel, its 150 to 200ms. On a self-hosted nodejs server, its 20-30ms. Same code

**Comment by balazsorban44** [comment]
Just guessing here as there was no reproduction provided, but it's probably related to https://github.com/nextauthjs/next-auth/issues/1535

So we introduced an API that does not even call `/api/auth/session`, resulting in much better performance on serverless platforms. See the docs: https://next-auth.js.org/configuration/nextjs#unstable_getserversession

Also, we are working on being framework agnostic, which means you'll be able to run `next-auth` on [Edge Functions](https://vercel.com/docs/concepts/functions/edge-functions), further reducing cold starts.

See #2294 and #5536

**Comment by avisra** [comment]
yea, its been difficult with the holidays - I still plan to do this reproduction though.

**Comment by ys-sherzad** [comment]
I think I have a similar problem A production deployed on **vercel** where a page using SSR to check if session exists or not and the `getServerSideProps` content is trivial where no other side-effects are taking place whatsoever is taking more **4 seconds** to load. I initially thought that this might be a cold start issue but after investigation I'm certain it's `getServerSession` pending that long:

<img width="1069" alt="Screenshot 2023-04-18 at 6 42 40 PM" src="https://user-images.githubusercontent.com/55835805/232846510-2de3751c-a3d3-47b2-aedd-ce72b3db8704.png">

```ts
export const getServerSideProps: GetServerSideProps = async (context) => {
  const session = await getServerSession(context.req, context.res, authOptions)
  if (session) {
    return {
      props: { user: session.user },
    }
  }
  return {
    redirect: {
      destination: `${baseUrl}/login`,
      permanent: false,
    },
  }
}
```
